x/emissions/abci_test.go

```diff
@@ -1,229 +1,180 @@
+	"github.com/zeta-chain/zetacore/testutil/sample"
+	emissionsModule "github.com/zeta-chain/zetacore/x/emissions"
+	emissionstypes "github.com/zeta-chain/zetacore/x/emissions/types"
+	observerTypes "github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func TestBeginBlocker(t *testing.T) {
+	t.Run("no observer distribution happens if emissions module account is empty", func(t *testing.T) {
+		k, ctx, _, zk := keepertest.EmissionsKeeper(t)
+		var ballotIdentifiers []string
+
+		observerSet := sample.ObserverSet(10)
+		zk.ObserverKeeper.SetObserverSet(ctx, observerSet)
+
+		ballotList := sample.BallotList(10, observerSet.ObserverList)
+		for _, ballot := range ballotList {
+			zk.ObserverKeeper.SetBallot(ctx, &ballot)
+			ballotIdentifiers = append(ballotIdentifiers, ballot.BallotIdentifier)
+		}
+		zk.ObserverKeeper.SetBallotList(ctx, &observerTypes.BallotListForHeight{
+			Height:           0,
+			BallotsIndexList: ballotIdentifiers,
+		})
+		for i := 0; i < 100; i++ {
+			emissionsModule.BeginBlocker(ctx, *k)
+			ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+		}
+		for _, observer := range observerSet.ObserverList {
+			_, found := k.GetWithdrawableEmission(ctx, observer)
+			require.False(t, found)
+		}
+	})
+	t.Run("no validator distribution happens if emissions module account is empty", func(t *testing.T) {
+		k, ctx, sk, _ := keepertest.EmissionsKeeper(t)
+		feeCollectorAddress := sk.AuthKeeper.GetModuleAccount(ctx, types.FeeCollectorName).GetAddress()
+		for i := 0; i < 100; i++ {
+			emissionsModule.BeginBlocker(ctx, *k)
+			ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+		}
+		require.True(t, sk.BankKeeper.GetBalance(ctx, feeCollectorAddress, config.BaseDenom).Amount.IsZero())
+	})
+	t.Run("tmp ctx is not committed if any of the distribution fails", func(t *testing.T) {
+		k, ctx, sk, _ := keepertest.EmissionsKeeper(t)
+		// Fund the emission pool to start the emission process
+		err := sk.BankKeeper.MintCoins(ctx, emissionstypes.ModuleName, sdk.NewCoins(sdk.NewCoin(config.BaseDenom, sdk.NewInt(1000000000000))))
+		require.NoError(t, err)
+		// Setup module accounts for emission pools except for observer pool , so that the observer distribution fails
+		_ = sk.AuthKeeper.GetModuleAccount(ctx, emissionstypes.UndistributedTssRewardsPool).GetAddress()
+		feeCollectorAddress := sk.AuthKeeper.GetModuleAccount(ctx, types.FeeCollectorName).GetAddress()
+		_ = sk.AuthKeeper.GetModuleAccount(ctx, emissionstypes.ModuleName).GetAddress()
+
+		for i := 0; i < 100; i++ {
+			// produce a block
+			emissionsModule.BeginBlocker(ctx, *k)
+			ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+		}
+		require.True(t, sk.BankKeeper.GetBalance(ctx, feeCollectorAddress, config.BaseDenom).Amount.IsZero())
+		require.True(t, sk.BankKeeper.GetBalance(ctx, emissionstypes.EmissionsModuleAddress, config.BaseDenom).Amount.Equal(sdk.NewInt(1000000000000)))
+	})
+}
+func TestObserverRewards(t *testing.T) {
```

Comment:
I think it would be better to have `TestDistributeObserverRewards` that directly test with the `DistributeObserverRewards` method to test rewards distribution.

`BeginBlocker` should test that `DistributeObserverRewards` is correctly called but the details and edge cases for observer distribution should be tested with `DistributeObserverRewards` directly